icu4c/source/i18n/repattrn.cpp

```diff
@@ -191,27 +193,38 @@ void RegexPattern::init() {
     fSets             = new UVector(fDeferredStatus);
     fInitialChars     = new UnicodeSet;
     fInitialChars8    = new Regex8BitSet;
-    fNamedCaptureMap  = uhash_open(uhash_hashUnicodeString,     // Key hash function
-                                   uhash_compareUnicodeString,  // Key comparator function
-                                   uhash_compareLong,           // Value comparator function
-                                   &fDeferredStatus);
     if (U_FAILURE(fDeferredStatus)) {
         return;
     }
     if (fCompiledPat == NULL  || fGroupMap == NULL || fSets == NULL ||
-            fInitialChars == NULL || fInitialChars8 == NULL || fNamedCaptureMap == NULL) {
+            fInitialChars == NULL || fInitialChars8 == NULL) {
         fDeferredStatus = U_MEMORY_ALLOCATION_ERROR;
         return;
     }
 
     // Slot zero of the vector of sets is reserved.  Fill it here.
     fSets->addElement((int32_t)0, fDeferredStatus);
+}
+
+
+bool RegexPattern::initNamedCaptureMap() {
+    if (fNamedCaptureMap) {
+        return true;
+    }
+    fNamedCaptureMap  = uhash_openSize(uhash_hashUnicodeString,     // Key hash function
+                                       uhash_compareUnicodeString,  // Key comparator function
+                                       uhash_compareLong,           // Value comparator function
+                                       7,                           // Initial table capacity
+                                       &fDeferredStatus);
+    if (U_FAILURE(fDeferredStatus)) {
+        return false;
+    }
 
     // fNamedCaptureMap owns its key strings, type (UnicodeString *)
     uhash_setKeyDeleter(fNamedCaptureMap, uprv_deleteUObject);
+    return true;
 }
 
-
 //--------------------------------------------------------------------------
 //
 //   zap            Delete everything owned by this RegexPattern.
```

Comment:
[Intend to comment line 262 in github UI]
nit: a style question. Although uhash_close() does check the pointer null before deletion. Should null check be added  before calling uhash_close()?

For example: RBBIRuleScanner::~RBBIRuleScanner() does check null before calling uhash_close().

Reply:
OK, Changed it for consistency with other related code. If I were doing it all from scratch, I think I would omit all the extra NULL checks.

Reply:
FYI I wouldn't add such NULL checks, just like we don't check for nullptr before `delete ptr;`.